sdk/cosmos/azure-cosmos-spark_3-0_2-12/src/main/scala/com/azure/cosmos/spark/PointWriter.scala

```diff
@@ -0,0 +1,152 @@
+// scalastyle:on underscore.import
+
+class PointWriter(container: CosmosAsyncContainer, cosmosWriteConfig: CosmosWriteConfig)
+  extends AsyncItemWriter
+    with CosmosLoggingTrait {
+
+
+  // TODO: moderakh executorService needs to get closed, otherwise we will have resource leak
+  private val executorService = Executors.newFixedThreadPool(cosmosWriteConfig.maxConcurrency)
+  private val capturedFailure = new AtomicReference[Throwable]()
+  private val pendingPointWrites = new TrieMap[Future[Unit], Boolean]()
+
+  def scheduleWrite(partitionKeyValue: PartitionKey, objectNode: ObjectNode): Unit = {
+    cosmosWriteConfig.itemWriteStrategy match {
+      case ItemWriteStrategy.ItemOverwrite => upsertWithRetryAsync(partitionKeyValue, objectNode)
+      case ItemWriteStrategy.ItemAppend => createWithRetryAsync(partitionKeyValue, objectNode)
+    }
+  }
+
+  def flushAndClose() {
+    for ((future, _) <- pendingPointWrites.snapshot()) {
+      Try(Await.result(future, Duration.Inf))
+    }
+  }
+
+  // scalastyle:off return
+  private def createWithRetryAsync(partitionKeyValue: PartitionKey,
+                                   objectNode: ObjectNode): Unit = {
+
+    val promise = Promise[Unit]()
+    pendingPointWrites.put(promise.future, true)
+
+    executeAsync(() => createWithRetry(partitionKeyValue, objectNode))
+      .onComplete {
+        case Success(value) => {
+          promise.success(Unit)
+          pendingPointWrites.remove(promise.future)
+        }
+        case Failure(e) => {
+          promise.failure(e)
+          capturedFailure.set(e)
+          pendingPointWrites.remove(promise.future)
+        }
+      }
+  }
+
+  private def upsertWithRetryAsync(partitionKeyValue: PartitionKey,
+                                   objectNode: ObjectNode): Unit = {
+    val promise = Promise[Unit]()
+    pendingPointWrites.put(promise.future, true)
+
+    executeAsync(() => upsertWithRetry(partitionKeyValue, objectNode))
+      .onComplete {
+        case Success(value) => {
+          promise.success(Unit)
+          pendingPointWrites.remove(promise.future)
+        }
+        case Failure(e) => {
+          promise.failure(e)
+          capturedFailure.set(e)
+          pendingPointWrites.remove(promise.future)
+        }
+      }
+  }
+
+  private def createWithRetry(partitionKeyValue: PartitionKey,
+                              objectNode: ObjectNode): Unit = {
+
+    var exceptionOpt = Option.empty[Exception]
+    for (attempt <- 1 to cosmosWriteConfig.maxRetryCount + 1) {
+      try {
+        // TODO: moderakh, there is room for further improvement by making this code nonblocking
+        // using reactive stream retry pattern
+        container.createItem(objectNode, partitionKeyValue, new CosmosItemRequestOptions()).block()
+        return
+      } catch {
+        case e: CosmosException if Exceptions.isResourceExistsException(e) =>
+          // TODO: what should we do on unique index violation? should we ignore or throw?
```

Comment:
My 2 cents - throw ( or ideally - after GA - write to poison message - HDFS location or similar -at least optionally) - that should ideally happen to any error where retries don't succeed

Reply:
for other non-retriable errors generally we throw. 

specifically regarding this comment, for Create Item we can't distinguish unique index violation form regular 409.

regular 409 are expected to be eaten and ignored on Create item but for unique index violation we probably should throw if we can distinsuish it from the regular 409.